### PR TITLE
[Improvement] Custom translation from Node IR

### PR DESCRIPTION
**Pull request by ayermolo**
*Description*:  Extending framework to support custom translation from Node IR to Instruction IR. Right now there is auto generated lowering for custom nodes to custom instruction IR. This allows more flexibility. For example regularly supported Node that can be lowered to something custom BE adds for Instruction IR

Refactored `IRGenVisitor` to a header so that it can be used by a backend to generate custom Instructions during IRGen. Otherwise no real change to IRGen or the `IRGenVisitor`. A couple helper functions were added too. And then there's an extra call to `Backend::generateInst()` at the beginning of `post()` to check if the backend wants to IRGen the current node, otherwise our default IRGen logic takes place.

*Testing*: Unit tests
*Documentation*:
fixes https://github.com/pytorch/glow/issues/2211

Please see a detailed explanation of how to fill out the fields in the relevant sections in PULL_REQUEST.md.

### CODE REVIEW

File: include/glow/IR/IRGen.h [hunk 1]
```diff
@@ -0,0 +1,58 @@
+#ifndef GLOW_IR_IRGEN_H
```
Comment (jfix71) [inline include/glow/IR/IRGen.h]: Please add the copyright header as seen in other `.h` files

File: include/glow/Optimizer/Optimizer.h [hunk 2]
```diff
@@ -58,7 +58,7 @@ Function *profileQuantization(Context &ctx, Function *F,
 
 /// Helper to generate and optimize IR from given Function \p F. \p
 /// shouldShareBuffers signifies whether to use the share buffers optimization.
```
Comment (jfix71) [inline include/glow/Optimizer/Optimizer.h]: Please update the doxygen here with `\p B`

File: include/glow/Backends/Backend.h [hunk 3]
```diff
@@ -85,6 +86,8 @@ class Backend {
 
   /// Optimize the Function \p F given compilation mode \p mode.
   void optimizeFunction(CompilationMode mode, Function *F);
+  /// \returns true if Backend generated Instruction
```
Comment (jfix71) [inline include/glow/Backends/Backend.h]: Please name the parameters and mention them here in the doxygen. Also end sentences with a period 😃.

File: include/glow/IR/IR.h [hunk 4]
```diff
@@ -276,7 +277,7 @@ class IRFunction final : public Named {
 
   /// Generate IR from the graph nodes. If the compilation mode is 'training'
   /// then this procedure will also generate the code for the backward pass.
```
Comment (jfix71) [inline include/glow/IR/IR.h]: Please update doxygen comments to mention `\p B`

File: lib/IR/IRGen.cpp [hunk 5]
```diff
@@ -14,23 +14,17 @@
...
 #include <unordered_set>
 
-using namespace glow;
-
 //===----------------------------------------------------------------------===//
 //              IRGen visitor - the code that generates the IR.
 //===----------------------------------------------------------------------===//
 
-namespace {
+namespace glow {
```
Comment (jfix71) [inline lib/IR/IRGen.cpp]: We prefer to use `using namespace glow;` in `cpp` files.

File: lib/IR/IRGen.cpp [hunk 6]
```diff
@@ -49,365 +43,345 @@ static size_t getConsecutiveSameNodeCount(NodeValueArrayRef inputs,
...
+void IRGenVisitor::post(Node *parent, Node *N) {
+  visited_.insert(N);
 
-    switch (N->getKind()) {
-    default:
-      llvm_unreachable("Unhandled node; perhaps the node should have been "
-                       "lowered, or the backend should have specified an IRGen "
-                       "case for this node to a backend-specific Instr.");
-      break;
+  // allows backend to generate their custom instrution IR
```
Comment (jfix71) [inline lib/IR/IRGen.cpp]: nit: Capitalize "allows" and end sentence with a period.

**Comment by nadavrot** [comment]
@ayermolo Thanks for doing this work. I have a few comments and questions:

It is very difficult for me to review IRGen.cpp because so much code in the file changes.

Could you give an example where custom IR translation is useful?  

Could you explain why the approach taken (IR visitor) is the right way to implement this functionality and not something else?

**Comment by ayermolo** [comment]
@nadavrot 
IRGen. Ah didn't notice. Hmm, looks like it's white spaces related. Locally beyondcompare ignores it. I'll fix on next commmit.

One usage model I envision where we have a common Node on Node IR level, and can take advantage of all the optimizations that apply to such nodes, but we want different Instruction IR then what is generated through common pass. One example is Concat. Right now it's lowered to bunch of insertTensors. Works for CPU, but might not for other custom back ends.

For this approach. One thing I tried to do is reduce the scope of changes. There was some refactoring or IRGen, but otherwise changes are minimal. Also in some sense it's similar to transform{Pre/Post}Lowering hooks in to backend. There is a common pass that goes through IRGEN, but now this allows Backend to intercept node being lowered in to instruction IR and do it's own thing.

I am certainly open to other approaches.

**Comment by jfix71** [comment]
I think the proposed approach here is reasonable and not too intrusive, and as @ayermolo said it's somewhat in line with approaches we've used elsewhere in the codebase. We are already using the IRGenVisitor here.

I could envision custom IRGen being useful -- I believe @opti-mix had previously mentioned desiring this support.

File: include/glow/IR/IRGen.h [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+#include <unordered_set>
+
+//===----------------------------------------------------------------------===//
+//              IRGen visitor - the code that generates the IR.
+//===----------------------------------------------------------------------===//
+
+namespace glow {
+
+/// A helper class for visiting and generating the dotty file from the graph.
```
Comment (jfix71) [inline include/glow/IR/IRGen.h]: This comment is wrong -- I see it was copied from IRGen.cpp, but can you fix it while you're moving the class too?

**Comment by ayermolo** [comment]
@nadavrot @jfix71 Thanks for input. I'll rework it and update this PR.

File: include/glow/IR/IR.h [hunk 8]
```diff
@@ -276,7 +277,8 @@ class IRFunction final : public Named {
...
-  void generateIR();
+  /// It allows Backend /p B to custom transalte from a Node to Instruction IR.
```
Comment (jfix71) [inline include/glow/IR/IR.h]: nit: typo `transalte`

File: include/glow/IR/IRGen.h [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+  /// map is used by instructions that want to inspect the generated
+  /// instructions. For example, gradient instructions that look at operands
+  /// that do not exist at the graph level. Not all variables are representible.
+  NodeToInstrTy nodeToInstr_;
+
+  /// The function that we are building.
+  IRFunction *F_;
+  /// The builder that adds instructions into the function.
+  IRBuilder builder_;
+  const Backend &B_;
```
Comment (jfix71) [inline include/glow/IR/IRGen.h]: Please add a doxygen comment for `B_`

File: tests/unittests/BackendTestUtils.h [hunk 10]
```diff
@@ -38,6 +45,61 @@ class MockBackend : public Backend {
...
   }
+
+  bool generateInst(Node *N, IRGenVisitor &irgen) const override {
+    return false;
+  }
+};
+
+/// MockBackend2 used only for unit testing to test custom lowering from Node to
+/// Instruction IR.
+class MockBackend2 : public Backend {
```
Comment (jfix71) [inline tests/unittests/BackendTestUtils.h]: nit: Can you name `MockBackend2` a little more informatively? E.g. `MockBackendCustomIRGen`

File: include/glow/IR/IRGen.h [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+
+  /// Saves the generated IR in \p v for the node \p N.
+  void registerIR(NodeValue N, Value *v);
+
+  /// Adds to Node \p N --> Instruction \p inst map.
+  void setNodeToIR(Node *N, Instruction *inst);
+
+  /// Return Instruction that is mapped to Node \p N.
+  /// If mapping doesn't exists returns nullptr.
+  Instruction *getNodeToIR(Node *N);
```
Comment (jfix71) [inline include/glow/IR/IRGen.h]: This doesn't appear to be used anywhere.

**Comment by opti-mix** [comment]
> My understanding is that it's mostly just refactoring `IRGenVisitor` to a header so that it can be used by a backend to generate custom Instructions during IRGen. This resulted in a whitespace change in `post()`, since the body of `post()` now has one less level of indentation.

> Otherwise I think it was no real change to IRGen or the `IRGenVisitor`. A couple helper functions were added too. And then there's an extra call to `Backend::generateInst()` at the beginning of `post()` to check if the backend wants to IRGen the current node, otherwise our default IRGen logic takes place.

This is also my understanding. I think it could be useful if the commit message and the PR description would clearly state, which part of the changes was just a re-factoring (e.g. moving `IRGenVisitor` to a header) and which part was the new functionality (e.g. helper functions and an extra call to `Backend::generateInst()`). This way it is easy for a reviewer to understand what's going on.

**Comment by nadavrot** [comment]
@ayermolo @jfix71 @opti-mix Okay. Sounds good. I don't have any other comments. Thank you.

File: include/glow/IR/IRGen.h (hunk 11 above)
Comment (ayermolo) [reply include/glow/IR/IRGen.h]: Well I figured since I added the setter function, add getter too in case anyone needs to use it. In IRGen.cpp map is used directly : nodeToInstr_[N] = V; or auto *PI = cast<MaxPoolWithXYInst>(nodeToInstr_[poolOut.getNode()]);.

I can remove it, and if someone else needs it they can add it.

**Comment by ayermolo** [comment]
@jfix71 @opti-mix Any other issues I need to resolve?
Want me to remove that un-used getter Jordan mentioned?

**Comment by opti-mix** [comment]
I'm OK with merging it. 

@jfix71 Do we want the changes to be squashed into a single commit before merging?

@nadavrot Could you formally approve the PR, because otherwise GH would not allow to merge it.

**Comment by ayermolo** [comment]
Great, thank  you very much.
